cpp/src/arrow/engine/CMakeLists.txt

```diff
@@ -58,7 +58,11 @@ foreach(LIB_TARGET ${ARROW_SUBSTRAIT_LIBRARIES})
   target_compile_definitions(${LIB_TARGET} PRIVATE ARROW_ENGINE_EXPORTING)
 endforeach()
 
-set(ARROW_SUBSTRAIT_TEST_LINK_LIBS ${ARROW_SUBSTRAIT_LINK_lIBS} ${ARROW_TEST_LINK_LIBS})
+if(NOT ARROW_SUBSTRAIT_LINK_LIBS)
+  set(ARROW_SUBSTRAIT_LINK_LIBS substrait)
```

Comment:
even-though this fixes the issue I am quite confused on why `list(APPEND ARROW_SUBSTRAIT_TEST_LINK_LIBS arrow_substrait_static)` is not containing `substrait` as a dependency to be linked statically as I would have expected. I've been able to reproduce also on `UBUNTU=22.04`.
@lidavidm what do you think? Any idea on a different approach I could take on this one?

Reply:
I think there's probably a missing dependency in ThirdpartyToolchain between the generated header and the command that generates it

Reply:
This is enough to reproduce it for me

```
cmake ../cpp -DCMAKE_BUILD_TYPE=Debug -GNinja -DARROW_BUILD_TESTS=ON -DARROW_BUILD_SHARED=OFF -DARROW_BUILD_STATIC=ON -DARROW_SUBSTRAIT=ON
ninja src/arrow/engine/CMakeFiles/arrow-substrait-substrait-test.dir/substrait/function_test.cc.o
```

Reply:
Maybe it has to go under `STATIC_INSTALL_INTERFACE_LIBS` instead?

Reply:
Thanks for taking a look @lidavidm!! I'll keep investigating as I don't think the current fix is good enough and we should find the root cause. Pinging @kou in case he can throw some light on what is going on.

Reply:
Could you try the following without your changes?

```diff
diff --git a/cpp/cmake_modules/BuildUtils.cmake b/cpp/cmake_modules/BuildUtils.cmake
index 888ca19af5..2b65fcc5f0 100644
--- a/cpp/cmake_modules/BuildUtils.cmake
+++ b/cpp/cmake_modules/BuildUtils.cmake
@@ -399,7 +399,7 @@ function(ADD_ARROW_LIB LIB_NAME)
     endif()
 
     if(ARG_STATIC_LINK_LIBS)
-      target_link_libraries(${LIB_NAME}_static LINK_PRIVATE
+      target_link_libraries(${LIB_NAME}_static PUBLIC
                             "$<BUILD_INTERFACE:${ARG_STATIC_LINK_LIBS}>")
       if(USE_OBJLIB)
         # Ensure that dependencies are built before compilation of objects in
```

Reply:
Thanks @kou this change also solves the issue.